src/services/ftp/backend.rs

```diff
@@ -0,0 +1,512 @@
+                .await
+                .map_err(|e| parse_io_error(e, "read", args.path()))?;
+        }
+
+        let r: BytesReader = match args.size() {
+            Some(size) => Box::new(reader.take(size)),
+            None => Box::new(reader),
+        };
+
+        Ok(r)
+    }
+
+    async fn write(&self, args: &OpWrite, r: BytesReader) -> Result<u64> {
+        let path = args.path();
+
+        let n = self.ftp_append(path, unshared_reader(r)).await?;
+
+        Ok(n)
+    }
+
+    async fn stat(&self, args: &OpStat) -> Result<ObjectMetadata> {
+        let path = args.path();
+
+        if path == self.root {
+            let mut meta = ObjectMetadata::default();
+            meta.set_mode(ObjectMode::DIR);
+            return Ok(meta);
+        }
+
+        let resp = self.ftp_stat(path).await?;
+
+        let mut meta = ObjectMetadata::default();
+
+        if resp.is_file() {
+            meta.set_mode(ObjectMode::FILE);
+        } else if resp.is_directory() {
+            meta.set_mode(ObjectMode::DIR);
+        } else {
+            meta.set_mode(ObjectMode::Unknown);
+        }
+
+        meta.set_content_length(resp.size() as u64);
+
+        meta.set_last_modified(OffsetDateTime::from(resp.modified()));
+
+        Ok(meta)
+    }
+
+    async fn delete(&self, args: &OpDelete) -> Result<()> {
+        let path = args.path();
+        if args.path().ends_with('/') {
+            self.ftp_rmdir(path).await?;
+        } else {
+            self.ftp_rm(path).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn list(&self, args: &OpList) -> Result<DirStreamer> {
+        let path = args.path();
+
+        let rd = self.ftp_list(path).await?;
+
+        Ok(Box::new(DirStream::new(Arc::new(self.clone()), path, rd)))
+    }
+}
+
+impl Backend {
+    pub(crate) fn ftp_connect(&self) -> Result<FtpStream> {
+        // connecting to remote address
+        let u = format! {"{}:{}", self.endpoint, self.port};
+
+        let mut ftp_stream = FtpStream::connect(&u)
+            .map_err(|e| new_request_connection_error("connection", &u, e))?;
+
+        // switch to secure mode if ssl/tls is on.
+        if self.tls {
+            ftp_stream = ftp_stream
+                .into_secure(TlsConnector::new().unwrap(), &u)
+                .map_err(|e| new_request_secure_error("connection", &u, e))?;
+        }
+
+        // login if needed
+        if !self.credential.0.is_empty() {
+            ftp_stream
+                .login(&self.credential.0, &self.credential.1)
+                .map_err(|e| new_request_sign_error("connection", &self.endpoint, e))?;
+        }
+
+        // change to the root path
+        ftp_stream
+            .cwd(&self.root)
+            .map_err(|e| new_request_root_eror("connection", &self.root, e))?;
+
+        Ok(ftp_stream)
+    }
+
+    pub(crate) async fn ftp_get(&self, path: &str) -> Result<Cursor<Vec<u8>>> {
+        let mut ftp_stream = self.ftp_connect()?;
```

Comment:
Do we need to reconnect to the FTP server for each operation? Building one or two TCP connections each time we read is kind of expensive I think.

Reply:
It indeed costs a lot resources, but it is the best I can do so far. Now I am trying to improve it by creating a connection pool based on [r2d2](https://github.com/sfackler/r2d2). It may take some time to get things done.

Reply:
It's OK to use a `Mutex` here. We can improve the concurrency in the future.